File: packages/cortex-dart/lib/src/resources/embeddings.dart
/// Embeddings resource.
library;

import '../errors.dart';
import '../http_client.dart';
import '../types.dart';

/// Provides access to the embeddings API.
class EmbeddingsResource {
  final CortexHttpClient _client;
  final String _baseUrl;
  final String? _defaultPool;

  /// Creates an [EmbeddingsResource].
  EmbeddingsResource({
    required CortexHttpClient client,
    required String baseUrl,
    String? defaultPool,
  })  : _client = client,
        _baseUrl = baseUrl,
        _defaultPool = defaultPool;

  /// Generates embeddings for the given input.
  ///
  /// [model] is the embedding model to use. If omitted, the pool's default model is used.
  /// [input] is a single string or list of strings to embed.
  /// [encodingFormat] optionally specifies the encoding format.
  /// [pool] overrides the client-level default pool for this request.
  Future<EmbeddingResponse> create({
    String? model,
    required dynamic input,
    String? encodingFormat,
    String? pool,
  }) async {
    if (model != null && model.isEmpty) {
      throw const CortexValidationException(
        message: 'Model must not be empty.',
        parameter: 'model',
      );
    }
    if (input is String && input.isEmpty) {
      throw const CortexValidationException(
        message: 'Input must not be empty.',
        parameter: 'input',
      );
    }
    if (input is List && input.isEmpty) {
      throw const CortexValidationException(
        message: 'Input list must not be empty.',
        parameter: 'input',
      );
    }

    final request = EmbeddingRequest(
      model: model,
      input: input,
      encodingFormat: encodingFormat,
    );

    // Resolve pool: per-request > client default > none
    final resolvedPool = pool ?? _defaultPool;
    final extraHeaders = <String, String>{};
    if (resolvedPool != null) {
      extraHeaders['x-cortex-pool'] = resolvedPool;
    }

    final json = await _client.post(
      '$_baseUrl/embeddings',
      body: request.toJson(),
      extraHeaders: extraHeaders.isNotEmpty ? extraHeaders : null,
    );
    return EmbeddingResponse.fromJson(json);
  }
}

File: packages/cortex-js/src/resources/embeddings.ts
import type { CortexClient } from '../client.js';
import type { EmbeddingCreateParams, EmbeddingResponse, RequestOptions } from '../types.js';
import { ValidationError } from '../errors.js';

export interface EmbeddingCreateOptions extends RequestOptions {
  /** Pool slug for routing this request. Overrides client-level defaultPool. */
  pool?: string;
}

export class Embeddings {
  private client: CortexClient;

  constructor(client: CortexClient) {
    this.client = client;
  }

  /**
   * Create embeddings for the given input.
   */
  async create(params: EmbeddingCreateParams, options?: EmbeddingCreateOptions): Promise<EmbeddingResponse> {
    if (params.model !== undefined && params.model === '') {
      throw new ValidationError('model must not be empty when provided', 'model');
    }
    if (!params.input) {
      throw new ValidationError('input is required', 'input');
    }
    if (Array.isArray(params.input) && params.input.length === 0) {
      throw new ValidationError('input must not be empty', 'input');
    }

    // Resolve pool: per-request > client default > none
    const pool = options?.pool ?? this.client._defaultPool;
    const poolHeaders: Record<string, string> = {};
    if (pool) {
      poolHeaders['x-cortex-pool'] = pool;
    }

    const requestOptions: RequestOptions = {
      ...options,
      headers: { ...poolHeaders, ...options?.headers },
    };

    const url = `${this.client._llmBaseUrl}/embeddings`;
    return this.client._request<EmbeddingResponse>('POST', url, {
      body: params,
      requestOptions,
    });
  }
}
